File: overlay/src/activity.test.ts
import { describe, it, expect } from 'vitest'
import { summarizeTx, ActivityProof } from './activity.js'

const NO_PROOFS: ActivityProof[] = []
const base = { txid: 't1', when: '2026-07-07T00:00:00.000Z', proofs: NO_PROOFS }

const inp = (identityKey: string, amount: number) => ({ identityKey, amount, assetId: 'a.0' })
const out = (outputIndex: number, identityKey: string, amount: number) => ({ outputIndex, identityKey, amount, assetId: 'a.0' })

describe('summarizeTx', () => {
  it('classifies a mint (no FT inputs) as issue to the largest output', () => {
    const e = summarizeTx({ ...base, ftInputs: [], ftOutputs: [out(0, 'alice', 100)] })
    expect(e).toMatchObject({ kind: 'issue', from: null, to: 'alice', amount: 100, assetId: 'a.0' })
  })

  it('classifies alice→bob with change back to alice as a transfer of the external amount', () => {
    const e = summarizeTx({
      ...base,
      ftInputs: [inp('alice', 100)],
      ftOutputs: [out(0, 'bob', 30), out(1, 'alice', 70)]
    })
    expect(e).toMatchObject({ kind: 'transfer', from: 'alice', to: 'bob', amount: 30 })
  })

  it('classifies alice→alice (all outputs self, conserved) as a 0-unit self transfer', () => {
    const e = summarizeTx({
      ...base,
      ftInputs: [inp('alice', 100)],
      ftOutputs: [out(0, 'alice', 40), out(1, 'alice', 60)]
    })
    expect(e).toMatchObject({ kind: 'self', from: 'alice', to: 'alice', amount: 0 })
  })

  it('classifies a burn (in > out, all change to self) as redeem of the difference', () => {
    const e = summarizeTx({
      ...base,
      ftInputs: [inp('alice', 100)],
      ftOutputs: [out(0, 'alice', 25)]
    })
    expect(e).toMatchObject({ kind: 'redeem', from: 'alice', to: null, amount: 75 })
  })

  it('classifies a full burn (no outputs) as redeem of the whole input', () => {
    const e = summarizeTx({ ...base, ftInputs: [inp('alice', 100)], ftOutputs: [] })
    expect(e).toMatchObject({ kind: 'redeem', from: 'alice', to: null, amount: 100 })
  })

  it('returns null for a tx with no FT movement at all', () => {
    expect(summarizeTx({ ...base, ftInputs: [], ftOutputs: [] })).toBeNull()
  })

  it('sums multiple external outputs and picks the largest as the recipient', () => {
    const e = summarizeTx({
      ...base,
      ftInputs: [inp('alice', 100)],
      ftOutputs: [out(0, 'bob', 10), out(1, 'carol', 50), out(2, 'alice', 40)]
    })
    expect(e).toMatchObject({ kind: 'transfer', to: 'carol', amount: 60 })
  })

  it('ignores unknown-owner outputs when deciding transfer vs self', () => {
    // Output with no verified linkage ('' identity) is not treated as an
    // external recipient — in=out and every known output is the sender's.
    const e = summarizeTx({
      ...base,
      ftInputs: [inp('alice', 100)],
      ftOutputs: [out(0, '', 30), out(1, 'alice', 70)]
    })
    expect(e).toMatchObject({ kind: 'self', amount: 0 })
  })
})

// ---------------------------------------------------------------------------
// buildActivity pagination — complete-group guarantee at page boundaries
// ---------------------------------------------------------------------------
import { buildActivity, LinkageRowLite, ActivityDeps } from './activity.js'

const link = (txid: string, outputIndex: number, identityKey: string, createdAt: string): LinkageRowLite => ({
  txid,
  outputIndex,
  identityKey,
  linkage: { prover: identityKey, verifier: 'v', counterparty: identityKey, keyID: `k-${txid}-${outputIndex}`, proofType: 1 },
  createdAt
})

/** Deps with no raw txs — groups become entries only if raw exists; here we
 *  only exercise the paging/grouping layer, so raw txs are irrelevant and
 *  every group is skipped, but the cursor math still runs. */
function pagingDeps (rows: LinkageRowLite[]): ActivityDeps {
  return {
    listLinkage: async (limit, before) => {
      let r = rows
      if (before != null) r = r.filter(x => new Date(x.createdAt).getTime() <= new Date(before).getTime())
      return r.slice(0, limit)
    },
    findLinkageByOutpoints: async () => [],
    findRawTxs: async () => new Map()
  }
}

describe('buildActivity pagination', () => {
  it('returns a null cursor when everything fits in one page', async () => {
    const rows = [link('t1', 0, 'a', '2026-07-07T10:00:00.000Z'), link('t2', 0, 'a', '2026-07-07T09:00:00.000Z')]
    const page = await buildActivity(pagingDeps(rows), { limit: 100 })
    expect(page.nextCursor).toBeNull()
  })

  it('drops the boundary-straddling group and points the cursor at it (inclusive)', async () => {
    // 10 single-output txs, newest first; page limit 2 → fetches 2+overlap rows,
    // sees more exist, drops the oldest fetched group and cursors to it.
    const rows = Array.from({ length: 20 }, (_, i) =>
      link(`t${i}`, 0, 'a', new Date(Date.UTC(2026, 6, 7, 10, 0, 59 - i)).toISOString()))
    const page = await buildActivity(pagingDeps(rows), { limit: 2 })
    expect(page.nextCursor).not.toBeNull()
    // Cursor is the createdAt of a fetched row — inclusive re-fetch re-serves
    // that row's whole tx group on the next page.
    expect(rows.some(r => new Date(r.createdAt).toISOString() === page.nextCursor)).toBe(true)
  })

  it('clamps limit into [1, 500]', async () => {
    let requested = 0
    const deps: ActivityDeps = {
      listLinkage: async (limit) => { requested = limit; return [] },
      findLinkageByOutpoints: async () => [],
      findRawTxs: async () => new Map()
    }
    await buildActivity(deps, { limit: 99999 })
    expect(requested).toBeLessThanOrEqual(508) // 500 + overlap
    await buildActivity(deps, { limit: 0 })
    expect(requested).toBeGreaterThanOrEqual(9) // 1 + overlap
  })
})

File: overlay-go/internal/activity/activity_test.go
package activity

// Task 17 tests: case-for-case port of overlay/src/activity.test.ts.
//
// TS test -> Go test mapping:
//   'classifies a mint (no FT inputs) as issue ...'          -> TestSummarizeTx_IssueMint
//   'classifies alice->bob with change ... transfer'         -> TestSummarizeTx_TransferWithChange
//   'classifies alice->alice (all outputs self) ...'         -> TestSummarizeTx_SelfZeroUnit
//   'classifies a burn (in > out, all change to self) ...'   -> TestSummarizeTx_RedeemPartialBurn
//   'classifies a full burn (no outputs) ...'                -> TestSummarizeTx_RedeemFullBurn
//   'returns null for a tx with no FT movement at all'       -> TestSummarizeTx_NoMovementReturnsNil
//   'sums multiple external outputs ...'                     -> TestSummarizeTx_MultipleExternalOutputsSumAndPickLargest
//   'ignores unknown-owner outputs ...'                      -> TestSummarizeTx_UnknownOwnerOutputIgnored
//
//   'returns a null cursor when everything fits in one page' -> TestBuildActivity_NullCursorSinglePage
//   'drops the boundary-straddling group ... (inclusive)'    -> TestBuildActivity_DropsBoundaryStraddlingGroup
//   'still cursors past a max-split transfer at limit 1 ...' -> TestBuildActivity_MaxSplitGroupAtLimitOne
//   'clamps limit into [1, 500]'                             -> TestBuildActivity_ClampsLimit
//
// Plus Go-specific: TestBuildActivity_ISOFormatHasMilliseconds (asserts the
// `when`/`nextCursor` wire format exactly matches JS's toISOString()).

import (
	"context"
	"fmt"
	"sort"
	"strings"
	"testing"
	"time"

	"github.com/sirdeggen/mandala/overlay-go/internal/mandala"
)

var noProofs = []Proof{}

func baseParams() SummarizeParams {
	return SummarizeParams{Txid: "t1", When: "2026-07-07T00:00:00.000Z", Proofs: noProofs}
}

func inp(identityKey string, amount int64) FtInput {
	return FtInput{IdentityKey: identityKey, Amount: amount, AssetID: "a.0"}
}

func out(outputIndex uint32, identityKey string, amount int64) FtOutput {
	return FtOutput{OutputIndex: outputIndex, IdentityKey: identityKey, Amount: amount, AssetID: "a.0"}
}

func strPtr(s string) *string { return &s }

func TestSummarizeTx_IssueMint(t *testing.T) {
	p := baseParams()
	p.FtInputs = nil
	p.FtOutputs = []FtOutput{out(0, "alice", 100)}
	e := SummarizeTx(p)
	if e == nil {
		t.Fatal("nil entry")
	}
	if e.Kind != "issue" || e.From != nil || e.To == nil || *e.To != "alice" || e.Amount != 100 || e.AssetID != "a.0" {
		t.Fatalf("entry = %+v", e)
	}
}

func TestSummarizeTx_TransferWithChange(t *testing.T) {
	p := baseParams()
	p.FtInputs = []FtInput{inp("alice", 100)}
	p.FtOutputs = []FtOutput{out(0, "bob", 30), out(1, "alice", 70)}
	e := SummarizeTx(p)
	if e == nil {
		t.Fatal("nil entry")
	}
	if e.Kind != "transfer" || e.From == nil || *e.From != "alice" || e.To == nil || *e.To != "bob" || e.Amount != 30 {
		t.Fatalf("entry = %+v", e)
	}
}

func TestSummarizeTx_SelfZeroUnit(t *testing.T) {
	p := baseParams()
	p.FtInputs = []FtInput{inp("alice", 100)}
	p.FtOutputs = []FtOutput{out(0, "alice", 40), out(1, "alice", 60)}
	e := SummarizeTx(p)
	if e == nil {
		t.Fatal("nil entry")
	}
	if e.Kind != "self" || e.From == nil || *e.From != "alice" || e.To == nil || *e.To != "alice" || e.Amount != 0 {
		t.Fatalf("entry = %+v", e)
	}
}

func TestSummarizeTx_RedeemPartialBurn(t *testing.T) {
	p := baseParams()
	p.FtInputs = []FtInput{inp("alice", 100)}
	p.FtOutputs = []FtOutput{out(0, "alice", 25)}
	e := SummarizeTx(p)
	if e == nil {
		t.Fatal("nil entry")
	}
	if e.Kind != "redeem" || e.From == nil || *e.From != "alice" || e.To != nil || e.Amount != 75 {
		t.Fatalf("entry = %+v", e)
	}
}

func TestSummarizeTx_RedeemFullBurn(t *testing.T) {
	p := baseParams()
	p.FtInputs = []FtInput{inp("alice", 100)}
	p.FtOutputs = nil
	e := SummarizeTx(p)
	if e == nil {
		t.Fatal("nil entry")
	}
	if e.Kind != "redeem" || e.From == nil || *e.From != "alice" || e.To != nil || e.Amount != 100 {
		t.Fatalf("entry = %+v", e)
	}
}

func TestSummarizeTx_NoMovementReturnsNil(t *testing.T) {
	p := baseParams()
	p.FtInputs = nil
	p.FtOutputs = nil
	if e := SummarizeTx(p); e != nil {
		t.Fatalf("entry = %+v, want nil", e)
	}
}

func TestSummarizeTx_MultipleExternalOutputsSumAndPickLargest(t *testing.T) {
	p := baseParams()
	p.FtInputs = []FtInput{inp("alice", 100)}
	p.FtOutputs = []FtOutput{out(0, "bob", 10), out(1, "carol", 50), out(2, "alice", 40)}
	e := SummarizeTx(p)
	if e == nil {
		t.Fatal("nil entry")
	}
	if e.Kind != "transfer" || e.To == nil || *e.To != "carol" || e.Amount != 60 {
		t.Fatalf("entry = %+v", e)
	}
}

func TestSummarizeTx_UnknownOwnerOutputIgnored(t *testing.T) {
	// Output with no verified linkage ('' identity) is not treated as an
	// external recipient — in=out and every known output is the sender's.
	p := baseParams()
	p.FtInputs = []FtInput{inp("alice", 100)}
	p.FtOutputs = []FtOutput{out(0, "", 30), out(1, "alice", 70)}
	e := SummarizeTx(p)
	if e == nil {
		t.Fatal("nil entry")
	}
	if e.Kind != "self" || e.Amount != 0 {
		t.Fatalf("entry = %+v", e)
	}
}

// ---------------------------------------------------------------------------
// Build pagination — complete-group guarantee at page boundaries
// ---------------------------------------------------------------------------

func link(txid string, outputIndex uint32, identityKey string, createdAt time.Time) mandala.LinkageRow {
	return mandala.LinkageRow{
		Txid:        txid,
		OutputIndex: outputIndex,
		IdentityKey: identityKey,
		Linkage: mandala.SpecificLinkage{
			Prover:       identityKey,
			Verifier:     "v",
			Counterparty: identityKey,
			KeyID:        fmt.Sprintf("k-%s-%d", txid, outputIndex),
			ProofType:    1,
		},
		CreatedAt: createdAt,
	}
}

// pagingDeps is the Go mirror of the TS test suite's pagingDeps: an
// in-memory fake exercising only the paging/grouping layer. findRawTxs
// always returns empty, so every group is skipped after grouping, but the
// cursor math still runs — exactly like the TS fixture's comment says.
func pagingDeps(rows []mandala.LinkageRow) Deps {
	return Deps{
		ListLinkage: func(_ context.Context, limit int64, before *time.Time) ([]mandala.LinkageRow, error) {
			r := rows
			if before != nil {
				filtered := make([]mandala.LinkageRow, 0, len(r))
				for _, x := range r {
					if !x.CreatedAt.After(*before) {
						filtered = append(filtered, x)
					}
				}
				r = filtered
			}
			if int64(len(r)) > limit {
				r = r[:limit]
			}
			return r, nil
		},
		FindLinkageByOutpoints: func(context.Context, []mandala.Outpoint) ([]mandala.LinkageRow, error) {
			return nil, nil
		},
		FindRawTxs: func(context.Context, []string) (map[string]string, error) {
			return map[string]string{}, nil
		},
	}
}

func TestBuildActivity_NullCursorSinglePage(t *testing.T) {
	rows := []mandala.LinkageRow{
		link("t1", 0, "a", mustParse(t, "2026-07-07T10:00:00.000Z")),
		link("t2", 0, "a", mustParse(t, "2026-07-07T09:00:00.000Z")),
	}
	page, err := Build(context.Background(), pagingDeps(rows), Opts{Limit: 100})
	if err != nil {
		t.Fatal(err)
	}
	if page.NextCursor != nil {
		t.Fatalf("nextCursor = %v, want nil", *page.NextCursor)
	}
}

func mustParse(t *testing.T, s string) time.Time {
	t.Helper()
	tm, err := time.Parse(time.RFC3339Nano, s)
	if err != nil {
		t.Fatal(err)
	}
	return tm
}

func TestBuildActivity_DropsBoundaryStraddlingGroup(t *testing.T) {
	// 20 single-output txs, newest first; page limit 2 -> fetches 2+overlap
	// rows, sees more exist, drops the oldest fetched group and cursors to it.
	rows := make([]mandala.LinkageRow, 20)
	for i := 0; i < 20; i++ {
		ts := time.Date(2026, 7, 7, 10, 0, 59-i, 0, time.UTC)
		rows[i] = link(fmt.Sprintf("t%d", i), 0, "a", ts)
	}
	page, err := Build(context.Background(), pagingDeps(rows), Opts{Limit: 2})
	if err != nil {
		t.Fatal(err)
	}
	if page.NextCursor == nil {
		t.Fatal("nextCursor = nil, want non-nil")
	}
	// Cursor is the createdAt of a fetched row -- inclusive re-fetch re-serves
	// that row's whole tx group on the next page.
	found := false
	for _, r := range rows {
		if isoMillis(r.CreatedAt) == *page.NextCursor {
			found = true
			break
		}
	}
	if !found {
		t.Fatalf("nextCursor %q does not match any fetched row's createdAt", *page.NextCursor)
	}
}

func TestBuildActivity_MaxSplitGroupAtLimitOne(t *testing.T) {
	// A transfer writes up to 9 output-linkage rows (recipient + 8 split
	// change). The overlap must cover a whole such group, or a limit-1 page
	// whose newest tx is a max-split transfer fills the entire fetch window
	// with one txid and pagination dies (hasMore true, but len(order) == 1).
	rows := make([]mandala.LinkageRow, 0, 11)
	big := time.Date(2026, 7, 7, 10, 0, 0, 0, time.UTC)
	for i := uint32(0); i < 9; i++ {
		rows = append(rows, link("big", i, "a", big))
	}
	rows = append(rows, link("older", 0, "a", time.Date(2026, 7, 7, 9, 0, 0, 0, time.UTC)))
	rows = append(rows, link("oldest", 0, "a", time.Date(2026, 7, 7, 8, 0, 0, 0, time.UTC)))

	page, err := Build(context.Background(), pagingDeps(rows), Opts{Limit: 1})
	if err != nil {
		t.Fatal(err)
	}
	if page.NextCursor == nil {
		t.Fatal("nextCursor = nil, want non-nil")
	}
}

func TestBuildActivity_ClampsLimit(t *testing.T) {
	var requested int64
	deps := Deps{
		ListLinkage: func(_ context.Context, limit int64, _ *time.Time) ([]mandala.LinkageRow, error) {
			requested = limit
			return nil, nil
		},
		FindLinkageByOutpoints: func(context.Context, []mandala.Outpoint) ([]mandala.LinkageRow, error) {
			return nil, nil
		},
		FindRawTxs: func(context.Context, []string) (map[string]string, error) {
			return map[string]string{}, nil
		},
	}

	if _, err := Build(context.Background(), deps, Opts{Limit: 99999}); err != nil {
		t.Fatal(err)
	}
	if requested > 509 { // 500 + overlap
		t.Fatalf("requested = %d, want <= 509", requested)
	}

	// Opts.Limit's zero value doubles as "unspecified" (the HTTP layer has no
	// way to distinguish an omitted query param from an explicit 0 -- see
	// mandala.Store.PageAdminHistory's identical divergence), so it maps to
	// the 100 default rather than TS's clamp-to-1. Limit: 1 exercises the
	// actual lower-bound clamp the TS test intends.
	if _, err := Build(context.Background(), deps, Opts{Limit: 1}); err != nil {
		t.Fatal(err)
	}
	if requested < 10 { // 1 + overlap
		t.Fatalf("requested = %d, want >= 10", requested)
	}
}

func TestBuildActivity_ClampsLimit_ZeroDefaultsTo100(t *testing.T) {
	var requested int64
	deps := Deps{
		ListLinkage: func(_ context.Context, limit int64, _ *time.Time) ([]mandala.LinkageRow, error) {
			requested = limit
			return nil, nil
		},
		FindLinkageByOutpoints: func(context.Context, []mandala.Outpoint) ([]mandala.LinkageRow, error) {
			return nil, nil
		},
		FindRawTxs: func(context.Context, []string) (map[string]string, error) {
			return map[string]string{}, nil
		},
	}
	if _, err := Build(context.Background(), deps, Opts{Limit: 0}); err != nil {
		t.Fatal(err)
	}
	if requested != 109 {
		t.Fatalf("requested = %d, want 109 (100 default + 9 overlap)", requested)
	}
}

func TestBuildActivity_ISOFormatHasMilliseconds(t *testing.T) {
	rows := []mandala.LinkageRow{
		link("t1", 0, "a", time.Date(2026, 7, 7, 10, 0, 0, 0, time.UTC)),
	}
	// Force a cursor by requiring more rows than available: limit 0 rows is
	// invalid, so instead directly check isoMillis's format contract.
	got := isoMillis(rows[0].CreatedAt)
	want := "2026-07-07T10:00:00.000Z"
	if got != want {
		t.Fatalf("isoMillis = %q, want %q", got, want)
	}
	if !strings.Contains(got, ".000Z") {
		t.Fatalf("isoMillis = %q, missing millisecond field", got)
	}
}

// Ensures sort.SliceStable ordering is used and doesn't panic on ties (no
// direct TS analogue -- guards the Go-specific stable-sort requirement noted
// in the entries.sort port).
func TestBuildActivity_StableSortOnEqualWhen(t *testing.T) {
	ts := time.Date(2026, 7, 7, 10, 0, 0, 0, time.UTC)
	entries := []Entry{
		{Txid: "a", When: isoMillis(ts)},
		{Txid: "b", When: isoMillis(ts)},
	}
	sort.SliceStable(entries, func(i, j int) bool { return entries[i].When > entries[j].When })
	if entries[0].Txid != "a" || entries[1].Txid != "b" {
		t.Fatalf("stable sort reordered ties: %+v", entries)
	}
}
